storage/innobase/log/log0recv.cc

```diff
@@ -3423,6 +3423,45 @@ void recv_reset_buffer() {
   recv_sys->recovered_offset = 0;
 }
 
+#ifdef XTRABACKUP
+static bool xb_double_bp(ulint *max_mem) {
+  if (!xtrabackup_use_free_memory_pct_set) return false;
+
+  /* find availabe free memory */
+  auto free_memory = xtrabackup::utils::free_memory();
+  xb::info() << "Maximum memory for buffer pool:" << free_memory;
+  xb::info() << "Current Buffer Pool Size: " << srv_buf_pool_size;
+
+  /* gradually increase the buffer pool until it can processed redo log in
+one iteration or it hit threshold */
+  if (srv_buf_pool_size * 2 <=
+      (free_memory * xtrabackup_use_free_memory_pct / 100)) {
+    dict_ind_init();
+    if (!dict_boot()) return false;
+    srv_buf_pool_size *= 2;
+
+    srv_buf_pool_size = buf_pool_size_align(srv_buf_pool_size);
+
+    xb::info() << "Increasing the buffer pool to " << srv_buf_pool_size;
+
+    buf_pool_resize();
+    recv_n_pool_free_frames *= 2;
+    *max_mem =
+        UNIV_PAGE_SIZE * (srv_buf_pool_size / UNIV_PAGE_SIZE -
+                          (recv_n_pool_free_frames * srv_buf_pool_instances));
+    dict_close();
+    ibuf_close();
+    dict_ind_init();
+    return true;
```

Comment:
In server, we only do  resize, why do we to close these dict, ibuf structures?

Reply:
The problem here is that on the server we have all those structures already initialized prior to allowing users to resize the buffer pool. Here we are on a stage pre DD initialization. We need to:
1) Initialize them prior to resize.
2) close them otherwise startup will fail when we reach the point where DD will be initialized.

We can either add a check at server startup code to validate if dic_sys is not a nullptr or adjust recv_recovery_begin to close it in case  xb_double_bp has initialized it. I would say the latter is better. 

I've added a new commit to adjust it

Reply:
@altmannmarcelo These are not big structs, I would rather initialize them on startup unconditionally and free at the end.